include/nbl/asset/utils/CVertexHashGrid.h

```diff
@@ -0,0 +1,212 @@
+	}
+
+	const collection_t& vertices() const { return m_vertices; }
+
+	collection_t& vertices(){ return m_vertices; }
+
+	inline uint32_t getVertexCount() const { return m_vertices.size(); }
+
+	template <typename Fn>
+	void iterateBroadphaseCandidates(const VertexData& vertex, Fn fn) const
+	{
+		std::array<uint32_t, 8> neighboringCells;
+		const auto cellCount = getNeighboringCellHashes(neighboringCells.data(), vertex);
+
+		//iterate among all neighboring cells
+		for (uint8_t i = 0; i < cellCount; i++)
+		{
+			const auto& neighborCell = neighboringCells[i];
+			BucketBounds bounds = getBucketBoundsByHash(neighborCell);
+			for (; bounds.begin != bounds.end; bounds.begin++)
+			{
+				const vertex_data_t& neighborVertex = *bounds.begin;
+				if (&vertex != &neighborVertex)
+					if (!fn(neighborVertex)) break;
+			}
+		}
+		
+	};
+
+private:
+	struct KeyAccessor
+	{
+		_NBL_STATIC_INLINE_CONSTEXPR size_t key_bit_count = 32ull;
+
+		template<auto bit_offset, auto radix_mask>
+		inline decltype(radix_mask) operator()(const VertexData& item) const
+		{
+			return static_cast<decltype(radix_mask)>(item.getHash() >> static_cast<uint32_t>(bit_offset)) & radix_mask;
+		}
+	};
+
+	static constexpr uint32_t primeNumber1 = 73856093;
+	static constexpr uint32_t primeNumber2 = 19349663;
+	static constexpr uint32_t primeNumber3 = 83492791;
+
+	static constexpr uint32_t invalidHash = 0xFFFFFFFF;
+
+	using sorter_t = std::variant<
+		core::LSBSorter<KeyAccessor::key_bit_count, uint16_t>,
+		core::LSBSorter<KeyAccessor::key_bit_count, uint32_t>,
+		core::LSBSorter<KeyAccessor::key_bit_count, size_t>>;
+	sorter_t m_sorter;
+
+	static sorter_t createSorter(size_t vertexCount)
+	{
+		if (vertexCount < (0x1ull << 16ull))
+			return core::LSBSorter<KeyAccessor::key_bit_count, uint16_t>();
+		if (vertexCount < (0x1ull << 32ull))
+			return core::LSBSorter<KeyAccessor::key_bit_count, uint32_t>();
+		return core::LSBSorter<KeyAccessor::key_bit_count, size_t>();
+	}
+
+	collection_t m_vertices;
+	const uint32_t m_hashTableMaxSize;
+	const float m_cellSize;
+
+	uint32_t hash(const VertexData& vertex) const
+	{
+		const hlsl::float32_t3 position = floor(vertex.getPosition() / m_cellSize);
+
+		return	((static_cast<uint32_t>(position.x) * primeNumber1) ^
+			(static_cast<uint32_t>(position.y) * primeNumber2) ^
+			(static_cast<uint32_t>(position.z) * primeNumber3))& (m_hashTableMaxSize - 1);
+	}
+
+	uint32_t hash(const hlsl::uint32_t3& position) const
+	{
+		return	((position.x * primeNumber1) ^
+			(position.y * primeNumber2) ^
+			(position.z * primeNumber3))& (m_hashTableMaxSize - 1);
+	}
+
+	uint8_t getNeighboringCellHashes(uint32_t* outNeighbors, const VertexData& vertex) const
+	{
+		hlsl::float32_t3 cellfloatcoord = floor(vertex.getPosition() / m_cellSize - hlsl::float32_t3(0.5));
+		hlsl::uint32_t3 baseCoord = hlsl::uint32_t3(static_cast<uint32_t>(cellfloatcoord.x), static_cast<uint32_t>(cellfloatcoord.y), static_cast<uint32_t>(cellfloatcoord.z));
```

Comment:
you might wanna summarize in comments and an ASCII diagram the logic I had to explain on discord